src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/DownloadManager.java

```diff
@@ -123,9 +128,14 @@ public Future<Path> startDownload(
       Path output,
       ExtendedEventHandler eventHandler,
       Map<String, String> clientEnv,
-      String context) {
-    return executorService.submit(
+      String context,
+      CountDownLatch doneSignal) {
+    Semaphore doneSignaller = new Semaphore(1);
```

Comment:
Instead of combining a semaphore and a latch, maybe we could use a single shared `Phaser` instead? A download can `register` at the start of the lambda and `arriveAndDeregister` in the `finally` block. Then `close` can wait on `arriveAndAwaitAdvance`. You can search for `Phaser` in the codebase to find another very similar usage.

Reply:
I don't see how to implement it using `Phaser`. We need three phases: Init (0), started (1) and done (2). Calling `awaitAdvance(1)` to wait for the done phase is not possible while in the init phase. @fmeum What is your idea?

Reply:
I may very well be missing something, but why do we need to track the `init` phase at all with a `Phaser`? If the task has never been scheduled by the executor, we also don't need to cancel it. If it has started, it will either have registered with the `Phaser` (and will thus be awaited with `awaitAndAdvance`) or not since it hasn't even reached the `register` call (you can check the return value to avoid registrations happening after the advance).

Just to clarify: With the `Phaser` approach, `addListener` also shouldn't be needed.

Reply:
Updated with `Phaser`, which is much better. Check the diff for the [two last commits together](https://github.com/bazelbuild/bazel/pull/23837/files/3b214852ac3a6ccf18b679dcfceb3e215efb5b26..8949c9898141542090724e244a4757d22f08faf8). Tested on my small reproduction but not in production.

Reply:
I think that we could even use a single `Phaser` per context, see https://github.com/fmeum/bazel/tree/23837-phaser. It doesn't give use the opportunity the log the time taken to wait for a particular download cancellation though.